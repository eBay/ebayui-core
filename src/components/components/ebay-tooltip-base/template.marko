<span
    w-on-expander-expand="handleExpand"
    w-on-expander-collapse="handleCollapse"
    w-body
    overlay-style=data.overlayStyle
    w-bind/>
