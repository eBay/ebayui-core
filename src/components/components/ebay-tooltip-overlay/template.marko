<script marko-init>
    var constants = require("./constants");
    var typeRoles = constants.typeRoles;
    var pointerStyles = constants.pointerStyles;
</script>

<var hasUserStyles=(
    data.styleTop ||
    data.styleLeft ||
    data.styleRight ||
    data.styleBottom
)/>
<var overlayStyles=(hasUserStyles ? {
    top: data.styleTop,
    left: data.styleLeft,
    right: data.styleRight,
    bottom: data.styleBottom
} : pointerStyles[data.pointer || "bottom"])/>

<div
    class="${data.type}__overlay"
    role=typeRoles[data.type]
    aria-labelledby=(data.type === "tourtip" && widget.elId("tourtip-label"))
    id=data.tooltipId
    style=overlayStyles
    w-bind>
    <span class="${data.type}__pointer ${data.type}__pointer--${data.pointer}"/>
    <div class="${data.type}__mask">
        <div class="${data.type}__cell">
            <div class="${data.type}__content">
                <if(data.heading)>
                    <h2
                        class=["${data.type}__heading", data.heading.class]
                        style=data.heading.style
                        w-id="tourtip-label"
                        w-body=data.heading.renderBody/>
                </if>

                <if(data.content)>
                    <div
                        body-only-if(!data.content.class && !data.content.style)
                        class=data.content.class
                        style=data.content.style
                        w-body=data.content.renderBody/>
                </if>
            </div>

            <if(data.type !== "tooltip")>
                <button
                    aria-label=data.a11yCloseText
                    class="${data.type}__close"
                    type="button"
                    w-on-click="handleCloseButton">
                    <span aria-hidden="true" role="img"/>
                </button>
            </if>
        </div>
    </div>
</div>
