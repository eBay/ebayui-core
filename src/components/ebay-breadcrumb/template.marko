<script marko-init>
    var processHtmlAttributes = require("../../common/html-attributes");
</script>

<nav w-bind
    aria-labelledby=widget.elId("breadcrumb-heading")
    class=["breadcrumb", data.class]
    style=data.style
    role="navigation"
    ${processHtmlAttributes(data)}>

    <${data.a11yHeadingTag || "h2"} id=widget.elId("breadcrumb-heading") class="clipped">
        ${data.a11yHeadingText}
    </>
    <ul>
        <li for(item in data.items | status-var="status")>
            <var current=(!item.href && status.isLast())/>
            <${item.href ? "a" : "button"}
                href=item.href
                aria-current=(current && "location")
                class=item.class
                style=item.style
                w-body=item.renderBody
                w-onclick=(!current && "handleClick")
                ${processHtmlAttributes(item)}/>
        </li>
    </ul>
</nav>
