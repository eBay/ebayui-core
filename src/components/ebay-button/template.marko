<script marko-init>
    var getWidgetId = require("../../common/get-marko-3-widget-id");
    var processHtmlAttributes = require("../../common/html-attributes");
</script>

<var size=(data.size || "medium")/>
<var priority=(data.priority || "secondary")/>
<var variant=(!data.variant && data.href ? "fake" : data.variant)/>
<var isIconVariant=(variant === 'icon')/>
<var isExpandVariant=(variant === 'expand')/>
<var isCtaVariant=(variant === 'cta')/>
<var isBadged=Boolean(data.badgeNumber && isIconVariant)/>
<var noText=(isIconVariant || isBadged || (isExpandVariant && data.noText))/>
<var baseClass=(variant ? "${variant}-btn" : "btn")/>
<var sizeClass="${baseClass}--${size}"/>
<var htmlAttributes=processHtmlAttributes(data)/>
<var tag=(data.href ? "a" : "button")/>

<${tag}
    ${htmlAttributes}
    w-bind
    w-onclick="handleClick"
    w-onkeydown="handleKeydown"
    id=(data.id || getWidgetId(out))
    class=[
        data.class,
        baseClass,
        noText && "${baseClass}--no-text",
        isBadged && "${baseClass}--badged",
        data.fluid && "${baseClass}--fluid",
        data.truncate && "${sizeClass}-truncated",
        data.fixedHeight && "${sizeClass}-fixed-height",
        !data.fixedHeight && !data.truncate && sizeClass,
        (priority === "secondary" || priority === "primary") && "${baseClass}--${priority}"
    ]
    style=data.style
    href=data.href
    type=(tag === "button" && data.type || "button")
    disabled=data.disabled
    aria-disabled=(data.partiallyDisabled && "true")>

    <if(isCtaVariant)>
        <span class="cta-btn__cell">
            <span w-body/>
            <ebay-icon type="inline" name="arrow-right-bold" class="cta-btn__icon" no-skin-classes/>
        </span>
    </if>
    <else>
        <var hasAriaLabel=Boolean(htmlAttributes.ariaLabel)/>
        <span body-only-if(!hasAriaLabel) w-body aria-hidden="true"/>
        <if(isBadged)>
            <ebay-badge
                number=data.badgeNumber
                type="icon"
                aria-label=(hasAriaLabel && data.badgeAriaLabel)
                aria-hidden=(hasAriaLabel && "true")/>
        </if>
    </else>
</>
