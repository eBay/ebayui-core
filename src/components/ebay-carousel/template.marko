<div class=data.classes style=data.style w-bind ${data.htmlAttributes}>
    <var config=data.config/>
    <var statusId=("carousel-status-" + widget.id)/>
    <div w-id="container" class=["carousel__container", data.bothControlsDisabled && "carousel__container--controls-disabled"]>
        <${data.a11yStatusTag}
            if(data.totalSlides >= 1)
            id=statusId
            class="clipped"
            aria-live=(data.autoplayInterval && !data.paused ? "off" : "polite")>
            <span>${data.a11yStatusText}</span>
        </>
        <button
            class="carousel__control carousel__control--prev"
            w-onclick=(!data.prevControlDisabled && "handleMove")
            data-direction=-1
            aria-describedby=statusId
            aria-label=data.a11yPreviousText
            aria-disabled=(data.prevControlDisabled && 'true')>
            <ebay-icon type="inline" name="chevron-left"/>
        </button>

        <ul w-id="list"
            class="carousel__list"
            style=(!config.nativeScrolling && data.offset && {
                transform:  "translate3d(" + (data.offset * -1) + "px,0,0)",
                transition: data.disableTransition ? 'none' : undefined
            })>
            <for(item in data.items)>
                <li ${item.htmlAttributes}
                    class=item.class
                    style=item.style
                    w-body=item.renderBody
                    aria-hidden=(!item.fullyVisible && 'true')
                    w-preserve-body-if(!!config.preserveItems)/>
            </for>
        </ul>

        <button
            w-id="next"
            class="carousel__control carousel__control--next"
            w-onclick=(!data.nextControlDisabled && "handleMove")
            data-direction=1
            aria-describedby=statusId
            aria-label=data.a11yNextText
            aria-disabled=(data.nextControlDisabled && 'true')>
            <ebay-icon type="inline" name="chevron-right"/>
        </button>

        <if(data.autoplayInterval && !data.bothControlsDisabled)>
            <if(data.paused)>
                <button
                    class="carousel__play"
                    aria-label=data.a11yPlayText
                    w-onclick="togglePlay">
                    <ebay-icon type="inline" name="play"/>
                </button>
            </if>
            <else>
                <button
                    class=`carousel__pause`
                    aria-label=data.a11yPauseText
                    w-onclick="togglePlay">
                    <ebay-icon type="inline" name="pause"/>
                </button>
            </else>
        </if>
    </div>

    <ul class="carousel__dots" if(!data.noDots && data.totalSlides > 1)>
        <li for(i from 0 to data.totalSlides - 1)>
            <var isActive=(i === data.slide)/>
            <button
                class={"carousel__dot--active": isActive}
                w-onclick=(!isActive && "handleDotClick")
                data-slide=i
                aria-describedby=statusId
                aria-disabled=(isActive && 'true')
                aria-label=(isActive ?
                    data.a11yCurrentText.replace('{currentSlide}', i + 1) :
                    data.a11yOtherText.replace('{slide}', i + 1))/>
        </li>
    </ul>
</div>
