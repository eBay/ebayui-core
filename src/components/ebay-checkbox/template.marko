<script marko-init>
    var processHtmlAttributes = require("../../common/html-attributes");
</script>

<span w-bind class=["checkbox", data.class] style=data.style>
    <input
        class="checkbox__control"
        type="checkbox"
        w-onchange="handleChange"
        disabled=data.disabled
        ${processHtmlAttributes(data)}/>
    <span class="checkbox__icon" hidden>
        <ebay-icon type="inline" name="checkbox-unchecked" class="checkbox__unchecked" focusable="false" no-skin-classes/>
        <ebay-icon type="inline" name="checkbox-checked" class="checkbox__checked" focusable="false" no-skin-classes/>
    </span>
</span>
