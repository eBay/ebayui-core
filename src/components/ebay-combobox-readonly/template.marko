<span
    class=data.class
    style=data.style
    ${data.htmlAttributes}
    w-on-expander-expand='handleExpand'
    w-on-expander-collapse='handleCollapse'
    w-preserve-attrs="class"
    w-bind>
    <span class=data.btnClass w-preserve-attrs="class">
        <input
            type='text'
            role="combobox"
            readonly
            value=(data.selectedOption && data.selectedOption.text)
            disabled=data.disabled
            aria-haspopup='listbox'
            aria-owns=widget.elId('combobox__options')
            aria-expanded='false'
            w-on-keydown='handleComboboxKeyDown'
            w-preserve-attrs="aria-expanded"
            ${data.htmlAttributes}/>
        <span class="combobox__icon" aria-hidden="true"></span>
    </span>
    <div
        if(data.options.length > 0)
        w-id='combobox__options'
        role='listbox'
        class=data.optionsClass>
        <div
            for(option in data.options)
            role='option'
            class=['combobox__option', option.class, (option.selected && 'combobox__option--active')]
            style=option.style
            ${option.htmlAttributes}
            tabindex="-1"
            w-on-click='handleOptionClick'
            w-preserve-attrs="tabindex,data-makeup-index"
            aria-selected=String(option.selected)
            data-option-value=option.value>
            <span>
                ${option.text}
            </span>
            <span class='combobox__status'/>
        </div>
    </div>
    <select
        hidden
        class='combobox__native'
        name=data.name>
        <option
            for(option in data.options)
            value=option.value
            selected=option.selected
            class=option.class>
        </option>
    </select>
</span>
