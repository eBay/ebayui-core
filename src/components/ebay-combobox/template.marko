<span
    class=data.class
    ${data.htmlAttributes}
    w-on-expander-expand='handleExpand'
    w-on-expander-collapse='handleCollapse'
    w-bind>
    <span class=data.btnClass w-preserve-attrs="class">
        <input
            type='text'
            value=(data.selectedOption && data.selectedOption.label)
            readonly
            disabled=data.disabled
            role='combobox'
            aria-haspopup='listbox'
            aria-owns=widget.elId('combobox__options')
            aria-expanded='false'
            w-on-keydown='handleComboboxKeyDown'
            w-preserve-attrs="aria-expanded"/>
        <span class="combobox__icon" aria-hidden="true"></span>
    </span>
    <div
        if(data.options.length > 0)
        w-id='combobox__options'
        role='listbox'
        class=data.optionsClass>
        <div
            for(option in data.options)
            role='option'
            class=['combobox__option', option.class]
            tabindex="-1"
            ${option.htmlAttributes}
            w-on-click='handleOptionClick'
            w-preserve-attrs="tabindex,data-makeup-index"
            aria-selected=String(option.selected)
            data-option-value=option.value>
            <span class='combobox__option-label'>
                ${option.label}
            </span>
            <span class='combobox__status'/>
        </div>
    </div>
    <select
        hidden
        class='combobox__native'
        name=data.name>
        <option
            for(option in data.options)
            value=option.value
            selected=option.selected
            class=option.class>
        </option>
    </select>
</span>
