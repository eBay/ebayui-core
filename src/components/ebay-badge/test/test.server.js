import { describe, it, expect } from "vitest";

import { render } from "@marko/testing-library";
import { testPassThroughAttributes } from "../../../common/test-utils/server";
import template from "../index.marko";

it("renders defaults", async () => {
    const { getByText } = await render(template, { number: 5 });
    expect(getByText(/\d+/)).toMatchSnapshot();
});

it("renders number with rounded-up value", async () => {
    const { getByText } = await render(template, { number: 5.6 });
    expect(getByText(/\d+/)).toMatchSnapshot();
});

it("does not render with negative value", async () => {
    const { queryByText } = await render(template, { number: -5 });
    expect(queryByText(/\d+/)).toMatchSnapshot();
});

describe("given number is a string", () => {
    it("renders number with coerced string", async () => {
        const { getByText } = await render(template, { number: "5" });
        expect(getByText(/\d+/)).toMatchSnapshot();
    });

    it("renders number with rounded-up string", async () => {
        const { getByText } = await render(template, { number: "5.4" });
        expect(getByText(/\d+/)).toMatchSnapshot();
    });

    it("does not renders with an invalid string", async () => {
        const { queryByText } = await render(template, { number: "five" });
        expect(queryByText(/\d+/)).toMatchSnapshot();
    });

    it("does not renders with a negative string", async () => {
        const { queryByText } = await render(template, { number: "-5" });
        expect(queryByText(/\d+/)).toMatchSnapshot();
    });
});

it("truncates when the value is greater than 99", async () => {
    const { getByText } = await render(template, { number: 150 });
    expect(getByText(/\d+/)).toMatchSnapshot();
});

testPassThroughAttributes(template, { input: { number: 1 } });
