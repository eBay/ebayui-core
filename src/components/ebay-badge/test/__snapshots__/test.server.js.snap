// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`does not render with negative value 1`] = `null`;

exports[`given number is a string > does not renders with a negative string 1`] = `null`;

exports[`given number is a string > does not renders with an invalid string 1`] = `null`;

exports[`given number is a string > renders number with coerced string 1`] = `
<span
  class="badge"
  role="img"
>
  5
</span>
`;

exports[`given number is a string > renders number with rounded-up string 1`] = `
<span
  class="badge"
  role="img"
>
  5
</span>
`;

exports[`passes through additional html attributes 1`] = `
<span
  class="badge class1"
  data-passed-through="true"
  data-testid="AttributePassthrough"
  role="img"
  style="color:red"
  type="number"
>
  1
</span>
`;

exports[`passes through additional html attributes 2`] = `
<span
  class="badge class1"
  data-passed-through="true"
  data-testid="AttributePassthrough"
  role="img"
  style="color:red"
  type="number"
>
  1
</span>
`;

exports[`renders defaults 1`] = `
<span
  class="badge"
  role="img"
>
  5
</span>
`;

exports[`renders number with rounded-up value 1`] = `
<span
  class="badge"
  role="img"
>
  6
</span>
`;

exports[`truncates when the value is greater than 99 1`] = `
<span
  class="badge"
  role="img"
>
  99+
</span>
`;
